Upstream AndroidX KlibDumpParser

fsladkey opened the pull request:
Add KlibDumpParser from AndroidX repository into util-klib-abi

As discussed with @fzhinkin, this is a slightly edited version of the KlibDumpParser from our repo. Most changes are in the tests to remove explicit references to androidx classes as well as some tests which parsed full dumps from androidx libraries because I wasn't sure that was appropriate to include. I left the original copyright but of course changed packages.

It's a large PR, I considered trying to break this up into pieces but I figured I'd start here and break it up if necessary. 

Looking forward to your feedback, thanks!

ivandev0 commented:
Hi @fsladkey!

Thank you for the contribution! I will kindly ask you to wait approximately a week until @fzhinkin is back from his vacation. He should be the main reviewer of this as he has the most experience in this area.

fsladkey commented:
Thanks @ivandev0, that sounds good to me!

fzhinkin reviewed (changes requested):
Sorry for a delay with the review. And thank you for contributing the parser, I really appreciate it!

There are a few implementation specific questions I have, but the major one is about separating parsing a merged dump file produced by BCV and parsing a dump for a single klib-target emitted by this module.
Otherwise, it looks good, nice job!

fsladkey commented:
Thanks for the review! All those comments make sense, I'll take a look shortly.

fzhinkin commented on compiler/util-klib-abi/src/org/jetbrains/kotlin/library/abi/parser/KlibParsingCursorExtensions.kt:
```diff
@@ -0,0 +1,518 @@
+private val classModifierRegex = Regex("^(inner|value|fun|open)")
+private val functionKindRegex = Regex("^(constructor|fun)")
+private val functionModifierRegex = Regex("^(inline|suspend)")
+private val packageNameRegex = Regex("^[a-zA-Z0-9.]+")
```
Could you please elaborate about the illegal character exception?
Package names like `schön.paket.name` or ``package with a space`` seems to work fine (well, at least Kotlin 2.2.0 successfully compile declarations from these packages for both JVM and K/N).

fzhinkin commented on compiler/util-klib-abi/test/org/jetbrains/kotlin/library/abi/parser/KLibDumpParserTest.kt:
```diff
@@ -0,0 +1,649 @@
+import org.junit.jupiter.api.Assertions.assertThrows
+import kotlin.toString
+
+class KlibDumpParserTest {
```
A simplest test that will rely on such an approach could look like:
```kotlin
abstract class AbstractKlibDumpParserTest {
    @OptIn(ExperimentalLibraryAbiReader::class)
    fun runTest(@TestDataFile filePath: String) {
        require(filePath.endsWith(".kt"))

        val dumpFileName = filePath.dropLast(2) + "klib_abi.txt" // but there are also some v1_sig dumps
        val rawDump = Paths.get(dumpFileName).readText()
        val parsedDump = KlibDumpParser(rawDump).parse()
        performComeCheck(parsedDump)
    }
}
```

and then you just need to mention it in `GenerateLibraryAbiReaderTests.kt`:
```kotlin
...
            testClass<AbstractKlibDumpParserTest> {
                model("content")
            }
...
```

In general, such generated tests are usually ment to be a compiler tests and use a sophisticated test infrastructure (https://github.com/JetBrains/kotlin/blob/master/compiler/test-infrastructure/ReadMe.md), but I doubt it would be useful for this particular case as we only need to parse dump, and don't have to run the whole compilation pipeline.

fsladkey commented:
Sorry for the delay once again. Fixed the issue with package names and also addressed some issues with formatting of context params. Let me know what you think!

fzhinkin commented on compiler/util-klib-abi/src/org/jetbrains/kotlin/library/abi/parser/KLibDumpParser.kt:
```diff
@@ -0,0 +1,342 @@
+import org.jetbrains.kotlin.library.abi.impl.AbiPropertyImpl
+import org.jetbrains.kotlin.library.abi.impl.AbiSignaturesImpl
+import org.jetbrains.kotlin.library.abi.impl.AbiTopLevelDeclarationsImpl
+import org.jetbrains.kotlin.library.abi.impl.AbiValueParameterImpl
```
```suggestion
import org.jetbrains.kotlin.library.abi.*
import org.jetbrains.kotlin.library.abi.impl.*
```

fzhinkin commented on compiler/util-klib-abi/src/org/jetbrains/kotlin/library/abi/parser/KLibDumpParser.kt:
```diff
@@ -0,0 +1,342 @@
+ * @property klibDump The text of the dump file
+ * @property filePath The file path to add to parse exceptions for clearer debugging (optional)
+ */
+class KlibDumpParser(klibDump: String, private val filePath: String? = null) {
```
Could you please annotate the class with `ExperimentalLibraryAbiReader`?
I thought about adding a separate annotation for it, but this is an ABI Reader, so the existing annotation seems to make sense here.

fzhinkin commented on compiler/util-klib-abi/src/org/jetbrains/kotlin/library/abi/parser/KlibParsingCursorExtensions.kt:
```diff
@@ -0,0 +1,526 @@
+private val varargSymbolRegex = Regex("^\\.\\.\\.")
+private val openParenRegex = Regex("^\\(")
+private val closeParenRegex = Regex("^\\)")
+private val reifiedRegex = Regex("reified")
```
```suggestion
private val reifiedRegex = Regex("^reified")
```

fzhinkin commented on compiler/util-klib-abi/src/org/jetbrains/kotlin/library/abi/parser/Cursor.kt:
```diff
@@ -0,0 +1,90 @@
+        peek: Boolean = false,
+        skipInlineWhitespace: Boolean = true,
+    ): String? {
+        val match = pattern.find(currentLine)
```
`currentLine` creates a substring on every *call* and this function uses it twice. It makes sense to cache its value locally.

fzhinkin replied:
It seems impossible with `Regex`, but Java's `Matcher` allows performing this "lines[rowIndex].startsWith(pattern, startIndex = currentColumn)" by setting a `region` and calling `lookingAt` (I was checking if it's possible to abstain from creating a substring here). It does not significantly affect the performance, though. :(

fzhinkin commented on compiler/util-klib-abi/src/org/jetbrains/kotlin/library/abi/parser/KlibParsingCursorExtensions.kt:
```diff
@@ -0,0 +1,526 @@
+    return parsed
+}
+
+internal fun Cursor.parseClassKind(peek: Boolean = false): AbiClassKind? {
```
Here (and in`Cursor.parsePropertyKind`), `peek` is always `false`, so the same change as above could be applied here too.

fzhinkin commented on compiler/util-klib-abi/src/org/jetbrains/kotlin/library/abi/parser/KlibParsingCursorExtensions.kt:
```diff
@@ -0,0 +1,526 @@
+private val closeParenRegex = Regex("^\\)")
+private val reifiedRegex = Regex("reified")
+private val contextRegex = Regex("^context")
+private val colonRegex = Regex("^:")
```
Don't get me wrong, I'm not suggesting to win a "1 billion klib dumps challenge" or something, but a few tweaks could be applied to the current implementation to keep it relatively simple, yet make it a bit faster.

Some of the regexes below don't have to be regexes at all: it's either a literal string, or set of literal strings; by introducing a `parseSymbol` overload capable of handling those (and not creating substring to check the current line), the performance could be improved.

fzhinkin commented on compiler/util-klib-abi/tests-gen/org/jetbrains/kotlin/library/abi/KlibDumpParserTestGenerated.java:
```diff
@@ -0,0 +1,187 @@
+/*
```
Now (9f259b975cedfc268784092a392340664d3b9ca6) these tests are automatically generated during the build and they no longer should be committed to the repo.

fsladkey commented:
Thanks for the review @fzhinkin! Good points. I think I've addressed them but let me know what you think.

fsladkey commented:
No problem @fzhinkin! Thanks for the review and great suggestions.

That sounds totally fine as long as it's in addition to the AOSP header!

fzhinkin commented:
Merged: https://github.com/JetBrains/kotlin/commit/f16880442c526cc3b13f9c5816c0e51bf54d792b

@fsladkey thank you for the contribution!
